File: Second Year/Graph Algorithms/Labs/PW1-C++/graph.cpp
//
// Created by Raluca on 30.03.2023.
//
// I am getting a few errors. What is wrong? I am using CLion. I am using C++17. I am using the latest version of CLion. 


#include <vector>
#include <unordered_map>
#include <algorithm>
#include <utility>
#include <unordered_set>

struct pair_hash {
    template <class T1, class T2>
    std::size_t operator() (const std::pair<T1, T2>& p) const {
        auto h1 = std::hash<T1>{}(p.first);
        auto h2 = std::hash<T2>{}(p.second);
        return h1 ^ h2;
    }
};


class TripleDictGraph {
public:
    TripleDictGraph(int number_of_vertices, int number_of_edges) :
            _number_of_vertices(number_of_vertices),
            _number_of_edges(number_of_edges)
    {
        for (int index = 0; index < number_of_vertices; ++index) {
            _dictionary_in[index] = {};
            _dictionary_out[index] = {};
        }
    }

    std::unordered_map<std::pair<int, int>, int> dictionary_cost() const {
        return _dictionary_cost;
    }

    std::unordered_map<int, std::vector<int>> dictionary_in() const {
        return _dictionary_in;
    }

    std::unordered_map<int, std::vector<int>> dictionary_out() const {
        return _dictionary_out;
    }

    int number_of_vertices() const {
        return _number_of_vertices;
    }

    int number_of_edges() const {
        return _number_of_edges;
    }

    std::vector<int> parse_vertices() const {
        std::vector<int> vertices;
        for (const auto& entry : _dictionary_in) {
            vertices.push_back(entry.first);
        }
        return vertices;
    }

    std::vector<int> parse_inbound(int x) const {
        return _dictionary_in.at(x);
    }

    std::vector<int> parse_outbound(int x) const {
        return _dictionary_out.at(x);
    }

    std::vector<std::pair<int, int>> parse_cost() const {
        std::vector<std::pair<int, int>> keys;
        for (const auto& entry : _dictionary_cost) {
            keys.push_back(entry.first);
        }
        return keys;
    }

    bool add_vertex(int x) {
        if (_dictionary_in.find(x) != _dictionary_in.end() &&
            _dictionary_out.find(x) != _dictionary_out.end()) {
            return false;
        }
        _dictionary_in[x] = {};
        _dictionary_out[x] = {};
        ++_number_of_vertices;
        return true;
    }

    bool remove_vertex(int x) {
        if (_dictionary_in.find(x) == _dictionary_in.end() &&
            _dictionary_out.find(x) == _dictionary_out.end()) {
            return false;
        }
        _dictionary_in.erase(x);
        _dictionary_out.erase(x);
        for (auto& entry : _dictionary_in) {
            entry.second.erase(std::remove(entry.second.begin(), entry.second.end(), x), entry.second.end());
        }
        for (auto& entry : _dictionary_out) {
            entry.second.erase(std::remove(entry.second.begin(), entry.second.end(), x), entry.second.end());
        }
        std::vector<std::pair<int, int>> keys_to_remove;
        for (const auto& entry : _dictionary_cost) {
            if (entry.first.first == x || entry.first.second == x) {
                keys_to_remove.push_back(entry.first);
                --_number_of_edges;
            }
        }
        for (const auto& key : keys_to_remove) {
            _dictionary_cost.erase(key);
        }
        --_number_of_vertices;
        return true;
    }

    bool add_edge(int x, int y, int cost) {
        if (_dictionary_in[y].count(x)) {
            return false;
        } else if (_dictionary_out[x].count(y)) {
            return false;
        } else if (_dictionary_cost.find(std::make_pair(x, y)) != _dictionary_cost.end()) {
            return false;
        }

        _dictionary_in[y].insert(x);
        _dictionary_out[x].insert(y);
        _dictionary_cost[std::make_pair(x, y)] = cost;
        _number_of_edges++;
        return true;
    }

    bool remove_edge(int x, int y) {
        if (_dictionary_in.find(y) == _dictionary_in.end() || _dictionary_in.find(x) == _dictionary_in.end() || _dictionary_out.find(y) == _dictionary_out.end() || _dictionary_out.find(x) == _dictionary_out.end()) {
            return false;
        } else if (!_dictionary_in[y].count(x)) {
            return false;
        } else if (!_dictionary_out[x].count(y)) {
            return false;
        } else if (_dictionary_cost.find(std::make_pair(x, y)) == _dictionary_cost.end()) {
            return false;
        }

        _dictionary_in[y].erase(x);
        _dictionary_out[x].erase(y);
        _dictionary_cost.erase(std::make_pair(x, y));
        _number_of_edges--;
        return true;
    }

    int in_degree(int x) const {
        if (_dictionary_in.find(x) == _dictionary_in.end()) {
            return -1;
        }
        return _dictionary_in.at(x).size();
    }

    int out_degree(int x) const {
        if (_dictionary_out.find(x) == _dictionary_out.end()) {
            return -1;
        }
        return _dictionary_out.at(x).size();
    }

    int find_if_edge(int x, int y) const {
        if (_dictionary_in[y].count(x)) {
            return _dictionary_cost.at(std::make_pair(x, y));
        } else if (_dictionary_out[x].count(y)) {
            return _dictionary_cost.at(std::make_pair(x, y));
        }
        return false;
    }

    bool change_cost(int x, int y, int cost) {
        if (_dictionary_cost.find(std::make_pair(x, y)) == _dictionary_cost.end()) {
            return false;
        }
        _dictionary_cost[std::make_pair(x, y)] = cost;
        return true;
    }

    TripleDictGraph make_copy() const {
        return TripleDictGraph(*this);
    }

private:
    int _number_of_vertices;
    int _number_of_edges;
    std::unordered_map<int, std::unordered_set<int>> _dictionary_in;
    std::unordered_map<int, std::unordered_set<int>> _dictionary_out;
    std::unordered_map<std::pair<int, int>, int, pair_hash> _dictionary_cost;
};

File: Second Year/Graph Algorithms/Labs/PW1/graph.py
import copy


class TripleDictGraph:
    def __init__(self, number_of_vertices, number_of_edges):
        self._number_of_vertices = number_of_vertices
        self._number_of_edges = number_of_edges
        self._dictionary_in = {}
        self._dictionary_out = {}
        self._dictionary_cost = {}
        for index in range(number_of_vertices):
            self._dictionary_in[index] = []
            self._dictionary_out[index] = []

    @property
    def dictionary_cost(self):
        return self._dictionary_cost

    @property
    def dictionary_in(self):
        return self._dictionary_in

    @property
    def dictionary_out(self):
        return self._dictionary_out

    @property
    def number_of_vertices(self):
        return self._number_of_vertices

    @property
    def number_of_edges(self):
        return self._number_of_edges

    def parse_vertices(self):
        vertices = list(self._dictionary_in.keys())
        for v in vertices:
            yield v

    def parse_inbound(self, x):
        for y in self._dictionary_in[x]:
            yield y

    def parse_outbound(self, x):
        for y in self._dictionary_out[x]:
            yield y

    def parse_cost(self):
        keys = list(self._dictionary_cost.keys())
        for key in keys:
            yield key

    def add_vertex(self, x):
        if x in self._dictionary_in.keys() and x in self._dictionary_out.keys():
            return False
        self._dictionary_in[x] = []
        self._dictionary_out[x] = []
        self._number_of_vertices += 1
        return True

    def remove_vertex(self, x):
        if x not in self._dictionary_in.keys() and x not in self._dictionary_out.keys():
            return False
        self._dictionary_in.pop(x)
        self._dictionary_out.pop(x)
        for key in self._dictionary_in.keys():
            if x in self._dictionary_in[key]:
                self._dictionary_in[key].remove(x)
            elif x in self._dictionary_out[key]:
                self._dictionary_out[key].remove(x)
        keys = list(self._dictionary_cost.keys())
        for key in keys:
            if key[0] == x or key[1] == x:
                self._dictionary_cost.pop(key)
                self._number_of_edges -= 1
        self._number_of_vertices -= 1
        return True

    def in_degree(self, x):
        if x not in self._dictionary_in.keys():
            return -1
        return len(self._dictionary_in[x])

    def out_degree(self, x):
        if x not in self._dictionary_out.keys():
            return -1
        return len(self._dictionary_out[x])

    def add_edge(self, x, y, cost):
        if x in self._dictionary_in[y]:
            return False
        elif y in self._dictionary_out[x]:
            return False
        elif (x, y) in self._dictionary_cost.keys():
            return False
        self._dictionary_in[y].append(x)
        self._dictionary_out[x].append(y)
        self._dictionary_cost[(x, y)] = cost
        self._number_of_edges += 1
        return True

    def remove_edge(self, x, y):
        if x not in self._dictionary_in.keys() or y not in self._dictionary_in.keys() or x not in self._dictionary_out.keys() or y not in self._dictionary_out.keys():
            return False
        if x not in self._dictionary_in[y]:
            return False
        elif y not in self._dictionary_out[x]:
            return False
        elif (x, y) not in self._dictionary_cost.keys():
            return False
        self._dictionary_in[y].remove(x)
        self._dictionary_out[x].remove(y)
        self._dictionary_cost.pop((x, y))
        self._number_of_edges -= 1
        return True

    def find_if_edge(self, x, y):
        if x in self._dictionary_in[y]:
            return self._dictionary_cost[(x, y)]
        elif y in self._dictionary_out[x]:
            return self._dictionary_cost[(x, y)]
        return False

    def change_cost(self, x, y, cost):
        if (x, y) not in self._dictionary_cost.keys():
            return False
        self._dictionary_cost[(x, y)] = cost
        return True

    def make_copy(self):
        return copy.deepcopy(self)


def write_graph_to_file(graph, file):
    file = open(file, "w")
    if graph.number_of_vertices > 0 and graph.number_of_edges:
        first_line = str(graph.number_of_vertices) + ' ' + str(graph.number_of_edges) + '\n'
        file.write(first_line)
        if len(graph.dictionary_cost) == 0 and len(graph.dictionary_in) == 0:
            raise ValueError("There is nothing that can be written!")
        for edge in graph.dictionary_cost.keys():
            new_line = "{} {} {}\n".format(edge[0], edge[1], graph.dictionary_cost[edge])
            file.write(new_line)
        for vertex in graph.dictionary_in.keys():
            if len(graph.dictionary_in[vertex]) == 0 and len(graph.dictionary_out[vertex]) == 0:
                new_line = "{}\n".format(vertex)
                file.write(new_line)
    else:
        first_line = 'We cannot create this graph' + '\n'
        file.write(first_line)
    file.close()


def read_graph_from_file(filename):
    file = open(filename, "r")
    line = file.readline()
    line = line.strip()
    vertices, edges = line.split(' ')
    graph = TripleDictGraph(int(vertices), int(edges))
    line = file.readline().strip()
    while len(line) > 0:
        line = line.split(' ')
        if len(line) == 1:
            graph.dictionary_in[int(line[0])] = []
            graph.dictionary_out[int(line[0])] = []
        elif len(line) == 3:
            graph.dictionary_in[int(line[1])].append(int(line[0]))
            graph.dictionary_out[int(line[0])].append(int(line[1]))
            graph.dictionary_cost[(int(line[0]), int(line[1]))] = int(line[2])
        line = file.readline().strip()
    file.close()
    return graph